src/c_kzg_4844.c

```diff
@@ -2875,18 +2721,24 @@ static C_KZG_RET recover_cells_impl(
 
     /* Identify missing cells */
```

Comment:
Your approach for populating `missing_cell_indices` is very readable, that's good 😄. 

But I wonder if the following is more efficient for large `num_cells` (and still readable): 
1. Create a bool array of length `CELLS_PER_EXT_BLOB` that contains `false` everwhere. It keeps track of which cells we have.
2. Iterate over `cell_indices` and mark the cells we have in the bool array as `true`.
3. Iterate over the bool array and add every cell for which we have `false` to `missing_cell_indices`.

If `num_cells` is almost `CELLS_PER_EXT_BLOB` (i.e., only a few cells missing), then your approach scales *quadratically* in `CELLS_PER_EXT_BLOB`, whereas what I suggest would scale *linearly*, something like `3 * CELLS_PER_EXT_BLOB`.  

Am I missing something? Did you test this? Or is it just that you did not want to use the long bool array? 😅

Reply:
Mentioned on telegram, will post here for prosperity -- for small lists, vectors have good cache utilization which is why I intitially went for it. I ran the below benchmark to check that I did not make a mistake and got ~0.04 milliseconds -- the bool array algorithm seems also quite simple, so I don't mind using that instead.


```c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <time.h>

#define ARRAY_SIZE 128

bool is_in_array(const uint64_t *arr, size_t arr_size, uint64_t value) {
    for (size_t i = 0; i < arr_size; i++) {
        if (arr[i] == value) {
            return true;
        }
    }
    return false;
}

int main() {
    uint64_t array[ARRAY_SIZE];
    uint64_t missing[ARRAY_SIZE];
    size_t missing_count = 0;

    // Initialize the array with some random values
    for (size_t i = 0; i < ARRAY_SIZE; i++) {
        array[i] = rand() % ARRAY_SIZE;
    }

    // Time the for loop
    clock_t start = clock();

    // Check for missing numbers, but make sure that we
    // search the whole array each time by using numbers greater
    // than ARRAY_SIZE.
    for (uint64_t i = ARRAY_SIZE + 1; i <= 2 * ARRAY_SIZE; i++) {
        if (!is_in_array(array, ARRAY_SIZE, i)) {
            missing[missing_count++] = i;
        }
    }

    clock_t end = clock();
    double cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;

    // Print the execution time
    printf("Execution time: %f seconds\n", cpu_time_used);

    return 0;
}
```

Reply:
ok, cool. Thanks for the benchmark.
Then I think we can just keep the `is_in_array`-code. :)